File: java/datastructures/LinkedList.java
// Linked List implementation in Java
// Singly linked list with append, prepend, delete, find, reverse, and utility methods

public class Node<T> {
    T data;
    Node<T> next;

    public Node(T data) {
        this.data = data;
        this.next = null;
    }
}

public class LinkedList<T> {
    private Node<T> head;
    private int size;

    public LinkedList() {
        this.head = null;
        this.size = 0;
    }

    public void append(T data) {
        Node<T> newNode = new Node<>(data);
        if (head == null) {
            head = newNode;
        } else {
            Node<T> current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
        size++;
    }

    public void prepend(T data) {
        Node<T> newNode = new Node<>(data);
        newNode.next = head;
        head = newNode;
        size++;
    }

    public boolean delete(T data) {
        if (head == null) return false;

        if (head.data.equals(data)) {
            head = head.next;
            size--;
            return true;
        }

        Node<T> current = head;
        while (current.next != null) {
            if (current.next.data.equals(data)) {
                current.next = current.next.next;
                size--;
                return true;
            }
            current = current.next;
        }
        return false;
    }

    public Node<T> find(T data) {
        Node<T> current = head;
        while (current != null) {
            if (current.data.equals(data)) {
                return current;
            }
            current = current.next;
        }
        return null;
    }

    public void reverse() {
        Node<T> prev = null;
        Node<T> current = head;
        Node<T> next = null;

        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
    }

    public boolean isEmpty() {
        return head == null;
    }

    public int getSize() {
        return size;
    }

    public T peek() {
        return head != null ? head.data : null;
    }

    public T getAt(int index) {
        if (index < 0 || index >= size) return null;

        Node<T> current = head;
        int count = 0;

        while (count < index) {
            current = current.next;
            count++;
        }

        return current.data;
    }

    public boolean insertAt(T data, int index) {
        if (index < 0 || index > size) return false;

        if (index == 0) {
            prepend(data);
            return true;
        }

        Node<T> newNode = new Node<>(data);
        Node<T> current = head;
        int count = 0;

        while (count < index - 1) {
            current = current.next;
            count++;
        }

        newNode.next = current.next;
        current.next = newNode;
        size++;
        return true;
    }

    public T removeAt(int index) {
        if (index < 0 || index >= size) return null;

        if (index == 0) {
            T data = head.data;
            head = head.next;
            size--;
            return data;
        }

        Node<T> current = head;
        int count = 0;

        while (count < index - 1) {
            current = current.next;
            count++;
        }

        T data = current.next.data;
        current.next = current.next.next;
        size--;
        return data;
    }

    public void clear() {
        head = null;
        size = 0;
    }

    public Object[] toArray() {
        Object[] result = new Object[size];
        Node<T> current = head;
        int index = 0;

        while (current != null) {
            result[index++] = current.data;
            current = current.next;
        }

        return result;
    }

    public void display() {
        System.out.println("LinkedList: " + java.util.Arrays.toString(toArray()));
    }
} 

File: python/datastructure/linked_list.py
# Linked List implementation in Python
# Singly linked list with append, prepend, delete, find, reverse, and utility methods

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.size = 0
    
    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
        self.size += 1
    
    def prepend(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
        self.size += 1
    
    def delete(self, data):
        if not self.head:
            return False
        
        if self.head.data == data:
            self.head = self.head.next
            self.size -= 1
            return True
        
        current = self.head
        while current.next:
            if current.next.data == data:
                current.next = current.next.next
                self.size -= 1
                return True
            current = current.next
        return False
    
    def find(self, data):
        current = self.head
        while current:
            if current.data == data:
                return current
            current = current.next
        return None
    
    def reverse(self):
        prev = None
        current = self.head
        next_node = None
        
        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        self.head = prev
    
    def is_empty(self):
        return self.head is None
    
    def get_size(self):
        return self.size
    
    def peek(self):
        return self.head.data if self.head else None
    
    def get_at(self, index):
        if index < 0 or index >= self.size:
            return None
        
        current = self.head
        count = 0
        
        while count < index:
            current = current.next
            count += 1
        
        return current.data
    
    def insert_at(self, data, index):
        if index < 0 or index > self.size:
            return False
        
        if index == 0:
            self.prepend(data)
            return True
        
        new_node = Node(data)
        current = self.head
        count = 0
        
        while count < index - 1:
            current = current.next
            count += 1
        
        new_node.next = current.next
        current.next = new_node
        self.size += 1
        return True
    
    def remove_at(self, index):
        if index < 0 or index >= self.size:
            return None
        
        if index == 0:
            data = self.head.data
            self.head = self.head.next
            self.size -= 1
            return data
        
        current = self.head
        count = 0
        
        while count < index - 1:
            current = current.next
            count += 1
        
        data = current.next.data
        current.next = current.next.next
        self.size -= 1
        return data
    
    def clear(self):
        self.head = None
        self.size = 0
    
    def to_array(self):
        result = []
        current = self.head
        
        while current:
            result.append(current.data)
            current = current.next
        
        return result
    
    def display(self):
        print(f"LinkedList: {self.to_array()}") 